flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/SchedulingPipelinedRegionComputeUtil.java

```diff
@@ -0,0 +1,186 @@
+
+        final List<Set<SchedulingExecutionVertex>> regionList =
+                new ArrayList<>(uniqueRegions(vertexToRegion));
+        final List<List<Integer>> outEdges =
+                buildOutEdgesDesc(vertexToRegion, regionList, executionVertexRetriever);
+        final Set<Set<Integer>> sccs =
+                StronglyConnectedComponentsComputeUtils.computeStronglyConnectedComponents(
+                        outEdges.size(), outEdges);
+
+        final Set<Set<SchedulingExecutionVertex>> mergedRegions =
+                Collections.newSetFromMap(new IdentityHashMap<>());
+        for (Set<Integer> scc : sccs) {
+            checkState(scc.size() > 0);
+
+            Set<SchedulingExecutionVertex> mergedRegion = new HashSet<>();
+            for (int regionIndex : scc) {
+                mergedRegion =
+                        mergeRegions(mergedRegion, regionList.get(regionIndex), vertexToRegion);
+            }
+            mergedRegions.add(mergedRegion);
+        }
+
+        return mergedRegions;
+    }
+
+    private static List<List<Integer>> buildOutEdgesDesc(
+            final Map<SchedulingExecutionVertex, Set<SchedulingExecutionVertex>> vertexToRegion,
+            final List<Set<SchedulingExecutionVertex>> regionList,
+            final Function<ExecutionVertexID, ? extends SchedulingExecutionVertex>
+                    executionVertexRetriever) {
+
+        final Map<Set<SchedulingExecutionVertex>, Integer> regionIndices = new IdentityHashMap<>();
+        for (int i = 0; i < regionList.size(); i++) {
+            regionIndices.put(regionList.get(i), i);
+        }
+
+        final List<List<Integer>> outEdges = new ArrayList<>(regionList.size());
+        for (Set<SchedulingExecutionVertex> currentRegion : regionList) {
+            final List<Integer> currentRegionOutEdges = new ArrayList<>();
+            for (SchedulingExecutionVertex vertex : currentRegion) {
+                for (SchedulingResultPartition producedResult : vertex.getProducedResults()) {
+                    if (producedResult.getResultType().isPipelined()) {
+                        continue;
+                    }
+                    for (ConsumerVertexGroup consumerVertexGroup :
+                            producedResult.getConsumerVertexGroups()) {
+                        for (ExecutionVertexID consumerVertexId : consumerVertexGroup) {
+                            SchedulingExecutionVertex consumerVertex =
+                                    executionVertexRetriever.apply(consumerVertexId);
+                            // Since we are merging SchedulingPipelinedRegions inside one
+                            // LogicalPipelinedRegion, if any vertex of the ConsumerVertexGroup
+                            // doesn't belong to this LogicalPipelinedRegion, we can just skip the
+                            // remaining vertices in this group. This can decrease the computation
+                            // complexity.
```

Comment:
This documentation is hard to understand because it introduced some outside assumptions.

Reply:
I tried to rewrite this comment.